sql/metrics/prometheus.go

```diff
@@ -0,0 +1,134 @@
+package metrics
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/sync/errgroup"
+	"time"
+
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/metrics"
+	"github.com/spacemeshos/go-spacemesh/sql"
+)
+
+const (
+	enabledDBStat = "ENABLE_DBSTAT_VTAB"
+	subsystem     = "database" // subsystem shared by all metrics exposed by this package.
+)
+
+// DBMetricsCollector collects metrics from db.
+type DBMetricsCollector struct {
+	logger        log.Logger
+	checkInterval time.Duration
+	db            *sql.Database
+	tablesList    map[string]struct{}
+	eg            errgroup.Group
+
+	tableSize *prometheus.GaugeVec
+	indexSize *prometheus.GaugeVec
+	totalSize *prometheus.GaugeVec
+}
+
+// NewDBMetricsCollector creates new DBMetricsCollector.
+func NewDBMetricsCollector(ctx context.Context, db *sql.Database, logger log.Logger, checkInterval time.Duration) {
+	collector := &DBMetricsCollector{
+		checkInterval: checkInterval,
+		logger:        logger.WithName("db_metrics"),
+		db:            db,
+		tableSize:     metrics.NewGauge("table_size", subsystem, "Size of table in bytes", []string{"name"}),
+		indexSize:     metrics.NewGauge("index_size", subsystem, "Size of index in bytes", []string{"name"}),
+		totalSize:     metrics.NewGauge("total_size", subsystem, "Total size of db in bytes", nil),
+	}
+	statEnabled, err := collector.checkCompiledWithDBStat()
+	if err != nil {
+		collector.logger.With().Error("error check compile options", log.Err(err))
+		return
+	}
+	if !statEnabled {
+		collector.logger.With().Info("sqlite compiled without `SQLITE_ENABLE_DBSTAT_VTAB`. Metrics will not collected")
+		return
+	}
+
+	collector.tablesList, err = collector.getListOfTables()
+	if err != nil {
+		collector.logger.With().Error("error get list of tables", log.Err(err))
+		return
+	}
+	collector.logger.With().Info("start collect stat")
+	collector.eg.Go(func() error {
```

Comment:
i think what @countvonzero meant is that it might be useful to wait for collector to finish. otherwise there are no reasons to use errgroup in here

Reply:
yes. we want to actively manage go routines to have a clean shutdown.